esp-hal/src/dma/buffers.rs

```diff
@@ -4,25 +4,334 @@ use core::{
         !matches!(self, Self::Disabled)
     }
+
+    // Size and address alignment as those come in pairs on current hardware.
+    const fn min_dram_alignment(self, direction: TransferDirection) -> usize {
+        if matches!(direction, TransferDirection::In) {
+            // NOTE(danielb): commenting this check is incorrect as per TRM, but works.
+            //                we'll need to restore this once peripherals can read a
+            //                different amount of data than what is configured in the
+            //                buffer.
+            // if cfg!(esp32) {
+            //     // NOTE: The size must be word-aligned.
+            //     // NOTE: The buffer address must be word-aligned
+            //     1
+            // }
+            if self.is_burst_enabled() {
+                // As described in "Accessing Internal Memory" paragraphs in the various TRMs.
+                4
+            } else {
+                1
+            }
+        } else {
+            // OUT transfers have no alignment requirements, except for ESP32, which is
+            // described below.
+            if cfg!(esp32) {
+                // SPI DMA: Burst transmission is supported. The data size for
+                // a single transfer must be four bytes aligned.
+                // I2S DMA: Burst transfer is supported. However, unlike the
+                // SPI DMA channels, the data size for a single transfer is
+                // one word, or four bytes.
+                4
+            } else {
+                1
+            }
+        }
+    }
+}
+
+const fn max(a: usize, b: usize) -> usize {
+    if a > b {
+        a
+    } else {
+        b
+    }
+}
+
+impl BurstConfig {
+    delegate::delegate! {
+        #[cfg(psram_dma)]
+        to self.internal_memory {
+            pub(super) const fn min_dram_alignment(self, direction: TransferDirection) -> usize;
+            pub(super) fn is_burst_enabled(self) -> bool;
+        }
+    }
+
+    /// Calculates an alignment that is compatible with the current burst
+    /// configuration.
+    ///
+    /// This is an over-estimation so that Descriptors can be safely used with
+    /// any DMA channel in any direction.
+    pub const fn min_compatible_alignment(self) -> usize {
+        let in_alignment = self.min_dram_alignment(TransferDirection::In);
+        let out_alignment = self.min_dram_alignment(TransferDirection::Out);
+        let alignment = max(in_alignment, out_alignment);
+
+        #[cfg(psram_dma)]
+        let alignment = max(alignment, self.external_memory as usize);
+
+        alignment
+    }
+
+    /// Calculates a chunk size that is compatible with the current burst
+    /// configuration's alignment requirements.
+    ///
+    /// This is an over-estimation so that Descriptors can be safely used with
+    /// any DMA channel in any direction.
+    pub const fn max_compatible_chunk_size(self) -> usize {
+        4096 - self.min_compatible_alignment()
```

Comment:
Nit: should we make this a const?

Reply:
What do you have in mind? The value depends on the device and burst settings

Reply:
I meant the `4096`, it's used inline in a few places

Reply:
yeah why not

Reply:
I've chosen to keep the literal in code, but deduplicate the calculation. Since 4096 by itself is max chunk size + 1, naming that would be either verbose, or inaccurate.